### ISSUE TITLE
ONNX shape inference does not infer shapes

### ISSUE THREAD

**Issue by andyzheng-snps** [onnx/onnx#4357]
Describe the bug
onnx.shape_inference.infer_shapes does not correctly infer shape of each layer.

System information
OS Platform and Distribution: linux 
ONNX version: 1.12.0
Python version: 3.9.7
pytorch version:1.11.0
Reproduction instructions
Describe the code to reproduce the behavior.
how to generate tranformer model:
import torch
transformer_model = torch.nn.Transformer(
    d_model=256,
    nhead=8,
    num_encoder_layers=1,
    num_decoder_layers=1,
    dim_feedforward=8,
    dropout=0.0,
    layer_norm_eps=1e-05
)

src = torch.rand((8, 48, 256))
tgt = torch.rand((8, 48, 256))
out = transformer_model(src, tgt)

torch.onnx.export(transformer_model,               # model being run
                  (src,tgt),                         # model input (or a tuple for multiple inputs)
                  "transformer_model.onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=15,          # the ONNX version to export the model to
                  do_constant_folding=False,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  dynamic_axes={'input' : {0 : 'batch_size'},    # variable length axes
                                'output' : {0 : 'batch_size'}})

----------infer shape
import onnx
from onnx.tools import update_model_dims
import numpy as np
import onnx.helper as helper
from onnx import shape_inference, TensorProto
import sys

ONNX_DTYPE = {
    0: TensorProto.FLOAT,
    1: TensorProto.FLOAT,
    2: TensorProto.UINT8,
    3: TensorProto.INT8,
    4: TensorProto.UINT16,
    5: TensorProto.INT16,
    6: TensorProto.INT32,
    7: TensorProto.INT64,
    8: TensorProto.STRING,
    9: TensorProto.BOOL
}

# load model
onnx_model = onnx.load("transformer_model.onnx")
graph = onnx_model.graph

# run node shape inference
node = graph.node
value_info = graph.value_info
print("Before shape inference: \n")
print(value_info)
print("------------------------------------------------------------")
print("After shape inference: \n")
inferred_onnx_model = shape_inference.infer_shapes(onnx_model)
onnx.checker.check_model(onnx_model)
inferred_graph = inferred_onnx_model.graph
inferred_value_info = inferred_graph.value_info
print(inferred_value_info)
onnx.save(inferred_onnx_model,"./new.onnx")

output:
..................
, name: "onnx::MatMul_370"
type {
  tensor_type {
    elem_type: 1
    shape {
      dim {
        dim_value: 8
      }
      dim {
        dim_value: 256
      }
    }
  }
}
, name: "onnx::Add_371"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "input.43"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "input.47"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Sub_374"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Pow_375"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Pow_376"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "onnx::ReduceMean_377"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_378"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_379"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "onnx::Sqrt_380"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Div_381"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Mul_382"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_383"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "input.51"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Sub_385"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Pow_386"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Pow_387"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "onnx::ReduceMean_388"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_389"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_390"
type {
  tensor_type {
    elem_type: 1
    shape {
    }
  }
}
, name: "onnx::Sqrt_391"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Div_392"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Mul_393"
type {
  tensor_type {
    elem_type: 1
  }
}
, name: "onnx::Add_394"
type {
  tensor_type {
    elem_type: 1
  }
}
]

**Comment by jcwchen** [comment]
Hi @andyzheng-snps,
Shape inference is not guaranteed to be complete. You can check the limitation [here](https://github.com/onnx/onnx/blob/main/docs/ShapeInference.md). It's possible that shape inference stops due to some dynamic situations.

**Comment by andyzheng-snps** [comment]
Hi jcwchen,

thanks for your quick response. So  I will use onnxruntime to get shape.

**Comment by wugoukanle** [comment]
there have another method，use onnx-tool to get shape in numpy
